Did task 1 and task 2
fixed task 3 by adding the exception, added a test to test the exception throwing
(re)did task 4 after finding out the current implementation was completely wrong, wrote some tests
did task 5, wrote some tests
did task 6, wrote all tests
